decisions/01-new-animation-api.md

```diff
@@ -0,0 +1,208 @@
+- `frameRate` (default: `60`) is a number that can be used to change the number of frames per second that the design function is called in `animation` mode.
+
+Since the new animation API encourages pure functions, the engines now ship with a `memo` function that can be used to cache things across function calls. Engine-specific helpers such as a `useCanvas` function will also be rewritten to create the canvas element on the first call and return it on following calls.
+
+The argument syntax is also changing slightly by placing the `frame` and `done` functions as root properties of the design function argument object. The `frame` function will only get passed to the design function in `customAnimation` mode.
+
+Here's a look at what this new animation API will look like for each engine.
+
+### `engine-canvas`
+
+```js
+export const handler = async ({ inputs, frameCount, done, useCanvas }) => {
+  const canvas = useCanvas(inputs.width, inputs.height);
+  if (frameCount >= 100) {
+    done(canvas);
+  }
+  // drawing code
+  return canvas;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-canvas'),
+  mode: 'animation'
+};
+```
+
+```js
+export const handler = async ({ inputs, frame, done, useCanvas }) => {
+  const canvas = useCanvas(inputs.width, inputs.height);
+  let frameCount = 0;
+  const myDrawLoop = () => {
+    frameCount++;
+    // drawing code
+    if (frameCount >= 100) {
+      done(canvas);
+    } else {
+      frame(canvas);
+      window.requestAnimationFrame(myDrawLoop);
+    }
+  };
+  myDrawLoop();
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-canvas'),
+  mode: 'customAnimation'
+};
+```
+
+### `engine-svg`
+
+```js
+export const handler = async ({ inputs, frameCount, done }) => {
+  // drawing code
+  if (frameCount >= 100) {
+    done(svgString);
+  }
+  return svgString;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-svg'),
+  mode: 'animation'
+};
+```
+
+```js
+export const handler = async ({ inputs, frame, done }) => {
+  let frameCount = 0;
+  const myDrawLoop = () => {
+    frameCount++;
+    // drawing code
+    if (frameCount >= 100) {
+      done(svgString);
+    } else {
+      frame(svgString);
+      window.requestAnimationFrame(myDrawLoop);
+    }
+  };
+  myDrawLoop();
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-svg'),
+  mode: 'customAnimation'
+};
+```
+
+### `engine-react`
+
+```js
+export const handler = async ({ inputs, done, frameCount }) => {
+  if (frameCount >= 100) {
+    done();
+  }
+  return <div></div>;
+};
+
+export const settings = {
+  engine: require('@mechanic-design/engine-react'),
+  mode: 'animation'
+};
+```
+
+The example above re-renders the component for every frame. If a user wants to use something like `react-spring` to speed up animation while keeping the frame-based approach, it could be done like this:
```

Comment:
Just leaving this here to consider when implementing this:

I think the way the react engine currently works this would cause a re-render of everything every time, because the engine would call React DOM's `render` over and over again. When implementing this using the the new animation approach the engine should probably be re-build to render a wrapping component only once and then use a drawLoop within that parent component to re-render to function over and over again.

Reply:
Yes, I agree. We need to take advantage of the React render loop, so the component should not render from scratch using `render` every time. good point!